### DISCUSSION TITLE
Help needed - New fields in transactions

### DISCUSSION THREAD

**Discussion by vincius0000** [enrique-lozano/Monekin#227] [General]
Hello!

I'm re-working Monekin to have it adapted to my needs, that includes:

1. Addition of new Text fields for Transactions
![image](https://github.com/user-attachments/assets/e0b1be39-2582-4a2b-8f01-6f4f389dd533)

2. Joint account option (ability for multiple people to use same account)
3. Online account (required for 2nd feature)

I need some help as I've managed to add fields, when entered they are saved in the database, but my issue is that when accessing `transaction` object then this field is null, even thought I see it when I inspect database manually

transaction_form.page.dart
```dart
// fillForm method, below setState method.
// [...]
    extReceiverNameController.text = transaction.extReceiverName ?? '';
    extReceiverAccountController.text = transaction.extReceiverAccount ?? '';
// [...]
}

  Widget buildExtReceiverNameField({bool isExpense = true}) {
    return ListTile(
      leading: const Icon(Icons.person_rounded),
      title: TextFormField(
        controller: extReceiverNameController,
        maxLength: maxLabelLenghtForDisplayNames,
        decoration: InputDecoration(
          isDense: false,
          filled: false,
          border: InputBorder.none,
          hintText: isExpense
              ? t.transaction.form.receiver_name
              : t.transaction.form.sender_name,
        ),
      ),
    );
  }

  Widget buildExtReceiverAccountField({bool isExpense = true}) {
    return ListTile(
      leading: const Icon(Icons.account_balance_rounded),
      title: TextFormField(
        controller: extReceiverAccountController,
        maxLength: maxLabelLenghtForDisplayNames,
        decoration: InputDecoration(
          isDense: false,
          filled: false,
          border: InputBorder.none,
          hintText: isExpense
              ? t.transaction.form.receiver_account
              : t.transaction.form.sender_account,
        ),
      ),
    );
  }
```

Transactions table schema additions
```sqlpl
  -- Name of external receiver/seller
  extReceiverName TEXT,

  -- Account of external receiver/seller
  extReceiverAccount TEXT,
```

Following added to `transaction.dart` constructor of `MoneyTransaction` class
```dart
      super.extReceiverName,
      super.extReceiverAccount,
```

Is there anything I'm missing?
Again, it saves well, but `transaction` object returns null when I try to fetch these fields as if they are null (while they are not).
Also, the reason I'm not doing it as a pull request to contribute to Monekin is that changes do have significant impact and online accounts aren't something Monekin supports now, this would only leave you with extra code that you wouldn't use.
Of course if Monekin planned on jumping to online accounts then I would help. This may be a separate project as well.

**Comment by enrique-lozano** [comment]
Hello @vincius0000! Thanks a lot for using the app. It's a pleasure to see that this is useful for a lot of persons. 

Regarding your code, probably you need to also add those fields in the getter function. The getters to get the transacción(s) are in one of the app services (since you close the conversation I suppose you found it).

I'm not really interested in adding those fields right now so thanks a lot for commenting here first instead of open a PR. The best part of an open source project is that the people can create their own versions of that project to fulfill their needs :)

Online support looks something big to me and I don't want to get involved in that at this moment. The first thing that I want to add in the near future is that the database file will be stored and automatically updated in your Google Drive account. That way the user don't have to manually save the database file every X days to do a backup.

Shared accounts also looks interesting to me but probably involves some paid subscription with any database external server (Amazon, Google Cloud...). I don't want to deal with any costs since the app is fully free and open source (I don't get any money from it).

Please keep me updated if you implement the online functionality since I'm really interested in your ideas. 

Good look creating your own version of Monekin ☺️

**Reply by vincius0000** [reply]
Hello @enrique-lozano !

Thank you for your answer!

> Regarding your code, probably you need to also add those fields in the getter function. The getters to get the transacción(s) are in one of the app services (since you close the conversation I suppose you found it).

It was something silly, it turns out I only needed to run the rebuild command (even thought I've used that command before, perhaps I didn't save one of the files in the editor while running the command). I didn't use migrations functionality as it's just me and my wife that's going to use it, so for now I'm keeping the code tidy.

> The first thing that I want to add in the near future is that the database file will be stored and automatically updated in your Google Drive account. That way the user don't have to manually save the database file every X days to do a backup.

That's a good idea! Could you create an issue of feature request type? I may find time for this, there will of course be a part where you will be needed, since oAuth has to be set up in order to gain access to Google drive.

> Shared accounts also looks interesting to me but probably involves some paid subscription with any database external server (Amazon, Google Cloud...). I don't want to deal with any costs since the app is fully free and open source (I don't get any money from it).

I understand, there's also security and liability, like, you don't want someone's personal information to leak out and get lawsuits for that. It's a bit easier for me as I'm not planning to publish re-created version of Monekin (due to higher cost & security risk). I'm planning to store the joint account transactions on a server (home or Amazon) and work on "event system", where every edit is saved as an event (ex. 5$ added to account on 2024-10-29) and during sync action the events are played back onto server. Might need to think of how to handle conflicts as well, but that's for later.

I've been thinking that since there's a lot of effort going to be put into this, it may be good to have it in open-source code, but perhaps only in the forked repository. Then technical people can re-use what I have but do it on their own server setup. The minimum I need to do for this to happen is create a documentation over what API calls the app makes and how to set up user authentication, additionally I could release very basic implementation of an API that technical people could simply copy and run. But that's for later.
